pkg/webhook/controlplane/kubeletconfigcodec.go

```diff
@@ -0,0 +1,120 @@
+// limitations under the License.
+
+package controlplane
+
+import (
+	"bytes"
+
+	"github.com/gardener/gardener-extensions/pkg/controller/operatingsystemconfig/oscommon/cloudinit"
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	kubeletconfigv1beta1 "k8s.io/kubelet/config/v1beta1"
+)
+
+// KubeletConfigCodec contains methods for encoding and decoding *kubeletconfigv1beta1.KubeletConfiguration objects
+// to and from *extensionsv1alpha1.FileContentInline.
+type KubeletConfigCodec interface {
+	// Encode encodes the given *kubeletconfigv1beta1.KubeletConfiguration into a *extensionsv1alpha1.FileContentInline.
+	Encode(*kubeletconfigv1beta1.KubeletConfiguration, *extensionsv1alpha1.FileContentInline) error
+	// Decode decodes a *kubeletconfigv1beta1.KubeletConfiguration from the given *extensionsv1alpha1.FileContentInline.
+	Decode(*extensionsv1alpha1.FileContentInline, *kubeletconfigv1beta1.KubeletConfiguration) error
+}
+
+// NewKubeletConfigCodec creates an returns a new KubeletConfigCodec.
+func NewKubeletConfigCodec() KubeletConfigCodec {
+	// Create scheme
+	scheme := runtime.NewScheme()
+	_ = kubeletconfigv1beta1.AddToScheme(scheme)
+
+	// Create codec for encoding / decoding to and from YAML
+	ser := json.NewYAMLSerializer(json.DefaultMetaFactory, scheme, scheme)
+	versions := schema.GroupVersions([]schema.GroupVersion{kubeletconfigv1beta1.SchemeGroupVersion})
+	codec := serializer.NewCodecFactory(scheme).CodecForVersions(ser, ser, versions, versions)
+
+	return &kubeletConfigCodec{
+		codec: codec,
+	}
+}
+
+type kubeletConfigCodec struct {
+	codec runtime.Codec
+}
+
+// Encode encodes the given *kubeletconfigv1beta1.KubeletConfiguration into a *extensionsv1alpha1.FileContentInline.
+func (c *kubeletConfigCodec) Encode(kubeletConfig *kubeletconfigv1beta1.KubeletConfiguration, fci *extensionsv1alpha1.FileContentInline) error {
+	// Encode kubelet configuration to YAML
+	var buffer bytes.Buffer
+	if err := c.codec.Encode(kubeletConfig, &buffer); err != nil {
+		return errors.Wrap(err, "could not encode kubelet configuration to YAML")
+	}
+	data := buffer.Bytes()
+
+	// Initialize file codec
+	fileCodec, err := getFileCodec(fci.Encoding)
+	if err != nil {
+		return err
+	}
+
+	// Encode data using the file codec, if needed
+	if fileCodec != nil {
+		if data, err = fileCodec.Encode(data); err != nil {
+			return errors.Wrap(err, "could not encode kubelet config file content data")
+		}
+	}
+
+	fci.Data = string(data)
+	return nil
+}
+
+// Decode decodes a *kubeletconfigv1beta1.KubeletConfiguration from the given *extensionsv1alpha1.FileContentInline.
+func (c *kubeletConfigCodec) Decode(fci *extensionsv1alpha1.FileContentInline, kubeletConfig *kubeletconfigv1beta1.KubeletConfiguration) error {
```

Comment:
Can we factor out the part that decodes the `FileContentInline` into a method like `DecodeFileContentInline`? Then we could also use the same in the `coreos-alicloud` controller.

Reply:
Factored out `FileContentInlineCodec` with `Encode` and `Decode`, still in `pkg/webhook/controlplane`. Should I move it somewhere else, e.g. `pkg/controller`?